### PR TITLE
[Merged by Bors] - feat(Probability/Kernel): disintegration of finite kernels

### PR DESCRIPTION
**Pull request by RemyDegenne**
Let `κ : kernel α (β × Ω)` be a finite kernel, where `Ω` is a standard Borel space. Then if `α` is countable or `β` has a countably generated σ-algebra (for example if it is standard Borel), then there exists a `kernel (α × β) Ω` called conditional kernel and denoted by `condKernel κ` such that `κ = fst κ ⊗ₖ condKernel κ`.

Properties of integrals involving `condKernel` are collated in the file `Integral.lean`.
The conditional kernel is unique (almost everywhere w.r.t. `fst κ`): this is proved in the file `Unique.lean`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by sgouezel** [comment]
When alpha is countable, can you explain why you need `MeasurableSingletonClass`? Otherwise, I would expect that everything is constant on the atoms of the partition generated by the sigma-algebra, and the quotient space is countable, so the general case should be reducable to the `MeasurableSingletonClass` case. I'm absolutely not saying that it should be implemented in this way with a quotient, rather that with this heuristics I don't see mathematically why `MeasurableSingletonClass` should be relevant.

**Comment by RemyDegenne** [comment]
The non-mathematical reason I added that hypothesis is that I used `measurable_from_prod_countable` (in the file `Probability.Kernel.Disintegration.Basic`). It asked for that typeclass and I added it without thinking about it.

**Comment by RemyDegenne** [comment]
I don't see how I could remove the `MeasurableSingletonClass` assumption without adding another assumption somewhere. To conclude that everything is constant on the atoms, I would need the codomain to have measurable singletons? Otherwise the function could take several values in a non-singleton atom, as long as all those values belong to some atom of the codomain?

Also the existence of a partition of atoms would mean that the measurable space is countably generated, but I think that's not necessarily the case even if the space is countable.

**Comment by sgouezel** [comment]
> I don't see how I could remove the `MeasurableSingletonClass` assumption without adding another assumption somewhere. To conclude that everything is constant on the atoms, I would need the codomain to have measurable singletons? Otherwise the function could take several values in a non-singleton atom, as long as all those values belong to some atom of the codomain?

I would think that everything is done taking preimages of measurable sets under measurable maps. Such preimages should be constant over atoms, right? (Warning: I haven't looked at the code, so I might be completely wrong)

> Also the existence of a partition of atoms would mean that the measurable space is countably generated, but I think that's not necessarily the case even if the space is countable.

I don't see the problem here. Define the atom of `x` as the intersection of all measurable sets that contain `x`. For any `y` not in the atom, one can take a measurable set containing `x` but not `y`, so the atom is in fact a countable intersection (of such sets over all `y` not belonging to the atom), and it belongs to the sigma-algebra. Then any measurable set is the (countable) union of the atoms it contains. But again, I don't think one should go through this argument for the construction, rather see where `MeasurableSingletonClass` is used and if one can remove it locally. If you want, I can have a look in the next days.

**Comment by RemyDegenne** [comment]
@sgouezel All prerequisites are now merged (thanks for the reviews!), and this is the final disintegration PR. I did not resolve your question about the `MeasurableSetSingleton` hypothesis: if you have time to have a look at the construction and see if you can remove that hypothesis, that would be great!

**Comment by sgouezel** [comment]
I've just pushed a commit removing `MeasurableSetSingleton`, following the sketch above. Can you tell me if you're happy with the modification? If so, I'll review the rest of the PR.

File: Mathlib/Probability/Kernel/Disintegration/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+  val p := (κ p.1).condKernel p.2
+  property := by
+    change Measurable ((fun q : β × α ↦ (κ q.2).condKernel q.1) ∘ Prod.swap)
+    refine (measurable_from_prod_countable' (fun a ↦ ?_) ?_).comp measurable_swap
+    · exact kernel.measurable (κ a).condKernel
+    · intro y y' x hy'
+      have : κ y' = κ y := by
+        ext s hs
+        exact mem_of_mem_measurableAtom hy'
+          (kernel.measurable_coe κ hs (measurableSet_singleton (κ y s))) rfl
```
Comment (RemyDegenne) [inline Mathlib/Probability/Kernel/Disintegration/Basic.lean:404]: @sgouezel That's the part of your explanation I did not quite get before. Now I see it. Thanks a lot!

**Comment by RemyDegenne** [comment]
Thanks for the help! Your code looks good. I saw a place where I think I can refactor my code to improve its clarity, so I'm turning that PR back to awaiting-author for now and will come back to you when it's again ready for review.

**Comment by RemyDegenne** [comment]
Actually I realize that I had already tried that refactor and it does not bring much benefits, so the PR is ready for review!

More explanations about that:
The definition `condKernelBorelSnd` takes a conditional CDF `f` (on the reals) and builds a kernel on a Borel space by going to the reals, building a kernel from the CDF (`hf.toKernel f`), then going back to the Borel space. It looks like I could first introduce another definition for the conditional kernel on R, then have `condKernelBorelSnd` do only the map and comap.
But actually `condKernelBorelSnd` does something slightly more subtle than map-then-comap and needs access to more properties of the kernel built from a conditional CDF (notably `ae_toKernel_eq_one` to show the conditional property), such that it is quite specific to the CDF construction and I might as well dispense with the intermediate definitions.

**Comment by sgouezel** [comment]
I'm confused by github here. I got an email saying that you had realized that the refactor wasn't worth it, and that the PR was ready for review, but I can't see your message on github and the PR is still awaiting author. Can you confirm that I did not dream this message, and that the PR is indeed ready for review?

**Comment by RemyDegenne** [comment]
The PR is not ready for review.
I wrote that message too fast, changed my mind once again about the refactor and then deleted the message. I forgot that github also sends an email.
The refactor is in progress. Sorry for the confusion!

**Comment by RemyDegenne** [comment]
The refactor is done and this is ready for review. The PR is big, but if I dismantle the old disintegration file I have to have all the replacements in the PR (which means I can't PR the files about integrals and uniqueness later).

File: Mathlib/Probability/Kernel/Disintegration/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+  conv_rhs => rw [this, ← hη']
+  exact compProd_fst_borelMarkovFromReal_eq_comapRight_compProd κ η hη
+
+end BorelSnd
+
+section CountablyGenerated
+
+open ProbabilityTheory.kernel
+
+lemma isRatCondKernelCDFAux_density_Iic (κ : kernel α (γ × ℝ)) [IsFiniteKernel κ] :
```
Comment (sgouezel) [inline Mathlib/Probability/Kernel/Disintegration/Basic.lean:241]: The next 70 lines are about kernels with values in a product where the second component is the real line, so they're interrupting a little bit the flow of the argument where you're moving things from the real line to a general standard Borel space. Would it make sense to move them to the top of the file (before the `BorelSnd` section)? Or to `CdfToKernel.lean`?

File: Mathlib/Probability/Kernel/Disintegration/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+end Unit
+
+section Measure
+
+variable {ρ : Measure (α × Ω)} [IsFiniteMeasure ρ]
+
+/-- Conditional kernel of a measure on a product space: a Markov kernel such that
+`ρ = ρ.fst ⊗ₘ ρ.condKernel` (see `MeasureTheory.Measure.compProd_fst_condKernel`). -/
+noncomputable
+def _root_.MeasureTheory.Measure.condKernel (ρ : Measure (α × Ω)) [IsFiniteMeasure ρ] :
```
Comment (sgouezel) [inline Mathlib/Probability/Kernel/Disintegration/Basic.lean:379]: `irreducible_def`, maybe, to hide the messy details of the construction (which should never be relevant once the API is set up)?

File: Mathlib/Probability/Kernel/Disintegration/Integral.lean [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+      = ∫⁻ x in Set.univ ×ˢ t, f x ∂ρ := by
+  rw [← set_lintegral_condKernel hf MeasurableSet.univ ht]; simp_rw [Measure.restrict_univ]
+#align probability_theory.set_lintegral_cond_kernel_univ_left MeasureTheory.Measure.set_lintegral_condKernel_univ_left
+
+end Lintegral
+
+section Integral
+
+variable {ρ : Measure (β × Ω)} [IsFiniteMeasure ρ]
+  {E : Type*} {f : β × Ω → E} [NormedAddCommGroup E] [NormedSpace ℝ E] [CompleteSpace E]
```
Comment (sgouezel) [inline Mathlib/Probability/Kernel/Disintegration/Integral.lean]: It seems to me that the `[CompleteSpace E]` assumption shouldn't be needed: if the space is not complete, then all integrals vanish by our choice of junk values, so all the theorems below remain true. I've taken the liberty to push a commit removing some of these assumptions here and there.

File: Mathlib/Probability/Kernel/Disintegration/Unique.lean [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+The conditional kernel of a measure is unique almost everywhere. -/
+
+/-- A s-finite kernel which satisfy the disintegration property of the given measure `ρ` is almost
+everywhere equal to the disintegration kernel of `ρ` when evaluated on a measurable set.
+
+This theorem in the case of finite kernels is weaker than `eq_condKernel_of_measure_eq_compProd`
+which asserts that the kernels are equal almost everywhere and not just on a given measurable
+set. -/
+theorem eq_condKernel_of_measure_eq_compProd' (κ : kernel α Ω) [IsSFiniteKernel κ]
```
Comment (sgouezel) [inline Mathlib/Probability/Kernel/Disintegration/Unique.lean:47]: I think I am unhappy with the assumptions in this theorem and in the next ones: if `ρ = ρ.fst ⊗ₘ κ` and `ρ` is a finite measure, then `κ a` should have finite mass almost everywhere wrt `ρ.fst`, and I feel that this should be sufficient to carry out the arguments in the lemma without the assumptions of s-finiteness or of finiteness. However, the lemmas were already there (at least in the measure version) and you just copied them, so this shouldn't hold the PR back. I'll investigate that once this one is merged, if you agree.

**Review by sgouezel — COMMENTED** [review]
This looks great, thanks! Sorry for the delay to review the PR (it means there is another merge conflict that you have to solve...)
bors d+

File: Mathlib/Probability/Kernel/Disintegration/Integral.lean (hunk 4 above)
Comment (RemyDegenne) [reply Mathlib/Probability/Kernel/Disintegration/Integral.lean]: Great, thanks!

**Comment by RemyDegenne** [comment]
Thanks for all the reviews and for your help on the disintegration PRs. No worries about the review time: I don't expect a quick review for a 1000 lines PR :)
I agree that the assumptions in the Unique file don't feel right. I won't work on that myself for now (at least not for the next days), so feel free to have a look.

bors r+

**Pull request closed without merging**